dev/src/dev/rewrite_defendpoint_1.clj

```diff
@@ -0,0 +1,422 @@
+
+;;;
+;;; rewrite body
+;;;
+
+(mu/defn- find-body :- ::zloc
+  [defendpoint :- ::zloc]
+  (z/right (find-argslist defendpoint)))
+
+(mu/defn- rewrite-body :- ::zloc
+  [_parsed     :- ::parsed
+   defendpoint :- ::zloc]
+  (let [body (find-body defendpoint)]
+    ;; if the first form is a map but there are more forms after, it's the schema map and we can remove it.
+    (if (and (= (z/tag body) :map)
+             (z/right body))
+      (z/remove body)
+      body)))
+
+;;;
+;;; rewrite entire form
+;;;
+
+(mu/defn- rewrite-defendpoint* :- ::zloc
+  [defendpoint :- ::zloc]
+  (let [parsed (parse-defendpoint-1-args defendpoint)]
+    (reduce
+     (fn [defendpoint f]
+       (z/subedit-node defendpoint (partial f parsed)))
+     defendpoint
+     [rewrite-defendpoint-symbol
+      rewrite-method
+      add-metadata-map
+      rewrite-args
+      rewrite-body])))
+
+(mu/defn- defendpoint-node? [node :- ::node]
+  (and
+   (= (n/tag node) :list)
+   (let [first-child (first (n/children node))]
+     (and
+      (n/symbol-node? first-child)
+      (= (n/sexpr first-child) 'api/defendpoint)))))
+
+(mu/defn- kondo-ignore-node? [node :- ::node]
+  (and
+   (= (n/tag node) :uneval)
+   (= (n/sexpr (first (n/children node)))
+      {:clj-kondo/ignore [:deprecated-var]})))
+
+(mu/defn- remove-preceding-kondo-ignore-node :- [:maybe ::zloc]
+  [defendpoint :- ::zloc]
+  (when-let [previous (z/left defendpoint)]
+    (when (some-> previous z/node kondo-ignore-node?)
+      (-> (z/subedit-node
+           (z/up defendpoint) ; move up to top-level <forms> node so the kondo ignore form is subeditable
+           (fn [_]
+             (z/remove previous)))
+          ; move back to first child
+          z/down))))
+
+(mu/defn- rewrite-defendpoint :- ::zloc
+  [defendpoint :- ::zloc]
+  (try
+    (-> defendpoint
+        rewrite-defendpoint*
+        remove-preceding-kondo-ignore-node)
+    (catch Throwable e
+      (println "Error rewriting defendpoint:" e)
+      defendpoint)))
+
+(mu/defn- rewrite-namespace :- ::zloc
+  [zloc :- ::zloc]
+  (if-let [defendpoint (z/find-next zloc #(defendpoint-node? (z/node %)))]
+    (let [zloc' (rewrite-defendpoint defendpoint)]
+      (recur zloc'))
+    zloc))
+
+;; TODO -- require needs to include api.macros.
+
+(defn rewrite-defendpoints! [^String filename & {:keys [write?], :or {write? true}}]
+  (println "Rewriting" filename)
+  (let [node (r.parser/parse-file-all filename)
+        zloc (-> (z/of-node node)
+                 rewrite-namespace)]
+    (letfn [(print-root [w]
+              (z/print-root zloc w))]
+      (if write?
+        (with-open [w (java.io.FileWriter. filename)]
+          (print-root w))
+        (print-root *out*)))))
+
+(def ^:private files
+  (->> (u.files/files-seq (u.files/get-path "src/metabase/api/"))
+       (map str)
+       (filter #(str/ends-with? % ".clj"))
+       sort))
+
+(defn rewrite-all! []
+  (doseq [file (take 3 files)]
+    (rewrite-defendpoints! file :write? true)))
```

Comment:
nit: This doesn't really rewrite-all! I see how this is the way to do it, but the name is a little misleading, maybe that's OK since this code will be removed once we've moved onto defendpoint2.

```suggestion
(defn rewrite-n! [n]
  (doseq [file (take n files)]
    (rewrite-defendpoints! file :write? true)))
```

Reply:
True, I'll fix it in subsequent PRs